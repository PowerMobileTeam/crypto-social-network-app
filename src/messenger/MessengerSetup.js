import React, {
  Component
} from 'react';

import {
  Text,
  View,
  TextInput,
  Alert,
  StyleSheet,
} from 'react-native';

import {
  inject,
  observer
} from 'mobx-react/native'

import { CommonStyle } from '../styles/Common';
import { ComponentsStyle } from '../styles/Components';
import Colors from '../styles/Colors';
import Button from '../common/components/Button';
import NavNextButton from '../common/components/NavNextButton';

/**
 * Messenger setup
 */
@inject('messengerList', 'user')
@observer
export default class MessengerSetup extends Component {

  /**
   * password
   * (don't use state to prevent render the component when password change)
   */
  password = '';
  confirm  = '';

  componentWillMount() {
    const { setParams } = this.props.navigation;
    let button;

    if (this.props.user.me.chat && !this.props.rekey) {
      button = (
        <NavNextButton
          onPress={this.unlock}
          title="UNLOCK"
          color={Colors.primary}
        />
      );
    } else {
      button = (
        <NavNextButton
          onPress={this.setup}
          title="SETUP"
          color={Colors.primary}
        />
      );
    }
    setParams({ headerRight: button });
  }

  unlock = () => {
    this.props.messengerList.getCryptoKeys(this.password)
      .then(resp => {
        this.handleOnDone(resp);
      });
  }

  setup = () => {
    if (this.password !== this.confirm) {
      Alert.alert('password and confirmation do not match!');
      return;
    }
    this.props.messengerList.doSetup(this.password)
      .then(resp => {
        this.handleOnDone(resp);
      });
  }

  handleOnDone(resp) {
    if (this.props.onDone) {
      this.props.onDone(resp);
    }
  }

  renderUnlock() {
    const unlocking = this.props.messengerList.unlocking;

    return (
      <View style={[CommonStyle.flexContainer, CommonStyle.padding2x, CommonStyle.backgroundLight]}>
        <View style={{ flexDirection: 'column', alignItems: 'stretch' }}>
          <TextInput
            style={ComponentsStyle.passwordinput}
            editable={true}
            underlineColorAndroid='transparent'
            placeholder='password...'
            secureTextEntry={true}
            onChangeText={(password) => this.password = password}
          />
        </View>

        <View style={{ paddingTop: 32 }}>
          <Text style={styles.infoText}>路 You only need to enter this encryption password once as long as you stay signed in.</Text>
          <Text style={styles.infoText}>路 It is important so that no one other than you and the people you are communicating with can access the content of your messages.</Text>
          <Text style={styles.infoText}>路 By default the content of your messages is fully encrypted. For heightened security you may wish to go to your app settings and turn off push notifications in order to disallow metadata from being tracked.</Text>
          <Text style={styles.infoText}>路 By default the content of your messages is fully encrypted. For heightened security you may wish to go to your app settings and turn off push notifications in order to disallow metadata from being tracked.</Text>
        </View>
      </View>
    )
  }

  renderOnboarding() {
    const unlocking = this.props.messengerList.unlocking;

    const text = this.props.user.me.chat ? 'Changing your encryption password will cause your previous messages to be unreadable':
      'Hey @'+this.props.user.me.name+'! It looks like you haven\'t setup your encrypted chat password yet. We recommend that you use a different password than your account password for added security.'

    return (
      <View style={[CommonStyle.flexContainer, CommonStyle.padding2x, CommonStyle.backgroundLight]}>
        <View style={{ flexDirection: 'column', alignItems: 'stretch' }}>
          <TextInput
            style={ComponentsStyle.passwordinput}
            editable={true}
            underlineColorAndroid='transparent'
            placeholder='password...'
            secureTextEntry={true}
            onChangeText={(password) => this.password = password}
          />
          <TextInput
            style={[ComponentsStyle.passwordinput, CommonStyle.marginTop2x]}
            editable={true}
            underlineColorAndroid='transparent'
            placeholder='confirm password...'
            secureTextEntry={true}
            onChangeText={(password) => this.confirm = password}
          />
        </View>

        <View style={{ paddingTop: 16 }}>
          <Text style={styles.infoText}>{text}</Text>
        </View>
      </View>
    );
  }

  /**
   * Render
   */
  render() {

    if (this.props.user.me.chat && !this.props.rekey) {
      return this.renderUnlock();
    } else {
      return this.renderOnboarding();
    }
  }
}

const styles = StyleSheet.create({
	infoText: {
    marginBottom: 16,
    color: '#BBB',
  },
});
