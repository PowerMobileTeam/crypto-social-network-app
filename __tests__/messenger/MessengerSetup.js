import 'react-native';
import React from 'react';
import {
  Alert
} from 'react-native';
import { shallow } from 'enzyme';
import MessengerSetup from '../../src/messenger/MessengerSetup';
import MessengerListStore from '../../src/messenger/MessengerListStore';
import UserStore from '../../src/auth/UserStore';
import appNavigation from '../../AppNavigation';
// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';

jest.mock('../../src/messenger/MessengerListStore');
jest.mock('../../src/auth/UserStore');
jest.mock('../../src/common/stores/NavigationStore');
jest.mock('../../src/common/components/NavNextButton');

Alert.alert = jest.fn();

/**
 * Tests
 */
describe('Messenger setup component', () => {
  let storem, userStore, navigation;


  beforeEach(() => {
    store = new MessengerListStore();
    userStore = new UserStore();
    navigation = appNavigation.buildNavigator();
    navigation.state.params = {};
    navigation.setParams = jest.fn();
    Alert.alert.mockClear();
  });

  it('should render correctly for setup', () => {
    const component = renderer.create(
      <MessengerSetup.wrappedComponent messengerList={store} user={userStore} navigation={navigation}/>
    ).toJSON();
    expect(component).toMatchSnapshot();
  });
  it('should render correctly for unlock', () => {
    userStore.me = {chat: true};
    const component = renderer.create(
      <MessengerSetup.wrappedComponent messengerList={store} user={userStore} navigation={navigation}/>
    ).toJSON();
    expect(component).toMatchSnapshot();
  });

  it('should show unlock button for users logged in messenger', () => {

    userStore.me = {chat: true};

    const wrapper = shallow(
      <MessengerSetup.wrappedComponent messengerList={store} user={userStore} navigation={navigation}/>
    );

    expect(navigation.setParams).toBeCalled();
    expect(navigation.setParams.mock.calls[0][0].headerRight.props.title).toEqual('UNLOCK');
  });

  it('should show setup button for users not logged in messenger', () => {

    userStore.me = {chat: false};

    const wrapper = shallow(
      <MessengerSetup.wrappedComponent messengerList={store} user={userStore} navigation={navigation}/>
    );

    expect(navigation.setParams).toBeCalled();
    expect(navigation.setParams.mock.calls[0][0].headerRight.props.title).toEqual('SETUP');
  });

  it('should show setup button for users not logged in messenger', () => {

    userStore.me = {chat: false};

    const wrapper = shallow(
      <MessengerSetup.wrappedComponent messengerList={store} user={userStore} navigation={navigation}/>
    );

    expect(navigation.setParams).toBeCalled();
    expect(navigation.setParams.mock.calls[0][0].headerRight.props.title).toEqual('SETUP');
  });

  it('should set the password property when input change', () => {

    userStore.me = {chat: false};

    const wrapper = shallow(
      <MessengerSetup.wrappedComponent messengerList={store} user={userStore} navigation={navigation}/>
    );

    // simulate user input
    const render = wrapper.dive();
    render.find('TextInput').forEach(child => {
      child.simulate('changeText', 'mypass');
    });

    expect(wrapper.instance().password).toEqual('mypass');
    expect(wrapper.instance().confirm).toEqual('mypass');
  });

  it('should warn the user if confirmation not match', () => {

    userStore.me = {chat: false};

    const wrapper = shallow(
      <MessengerSetup.wrappedComponent messengerList={store} user={userStore} navigation={navigation}/>
    );

    // simulate user input
    const render = wrapper.dive();
    render.find('TextInput').forEach((child, i) => {
      child.simulate('changeText', 'mypass'+i);
    });

    expect(wrapper.instance().password).toEqual('mypass0');
    expect(wrapper.instance().confirm).toEqual('mypass1');

    wrapper.instance().setup();

    expect(Alert.alert).toBeCalledWith('password and confirmation do not match!');
  });

  it('should call the method getCrytoKeys of the store with password', async () => {

    userStore.me = {chat: true};

    const onDone = jest.fn();

    const wrapper = shallow(
      <MessengerSetup.wrappedComponent messengerList={store} user={userStore} navigation={navigation} onDone={onDone}/>
    );

    store.getCryptoKeys.mockResolvedValue(true);

    // simulate user input
    const render = wrapper.dive();
    render.find('TextInput').forEach((child) => {
      child.simulate('changeText', 'mypass');
    });

    await wrapper.instance().unlock();

    // store should be called
    expect(store.getCryptoKeys).toBeCalledWith('mypass');
    // on done should be called
    expect(onDone).toBeCalled();
  });

  it('should call the method doSetup of the store with password', async () => {

    userStore.me = {chat: false};

    const onDone = jest.fn();

    const wrapper = shallow(
      <MessengerSetup.wrappedComponent messengerList={store} user={userStore} navigation={navigation} onDone={onDone}/>
    );

    store.doSetup.mockResolvedValue(true);

    // simulate user input
    const render = wrapper.dive();
    render.find('TextInput').forEach((child) => {
      child.simulate('changeText', 'mypass');
    });

    await wrapper.instance().setup();

    // store should be called
    expect(store.doSetup).toBeCalledWith('mypass');
    // on done should be called
    expect(onDone).toBeCalled();
  });
});