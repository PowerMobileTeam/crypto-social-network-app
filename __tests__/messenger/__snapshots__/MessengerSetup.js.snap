// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Messenger setup component should render correctly for setup 1`] = `
<View
  style={
    Array [
      Object {
        "flex": 1,
      },
      Object {
        "padding": 10,
      },
      Object {
        "backgroundColor": "#f7f7f7",
      },
    ]
  }
>
  <View
    style={
      Object {
        "alignItems": "stretch",
        "flexDirection": "column",
      }
    }
  >
    <TextInput
      allowFontScaling={true}
      editable={true}
      onChangeText={[Function]}
      placeholder="password..."
      secureTextEntry={true}
      style={
        Object {
          "backgroundColor": "#FFF",
          "borderColor": "#ECECEC",
          "borderRadius": 3,
          "borderWidth": 1,
          "padding": 16,
        }
      }
      underlineColorAndroid="transparent"
    />
    <TextInput
      allowFontScaling={true}
      editable={true}
      onChangeText={[Function]}
      placeholder="confirm password..."
      secureTextEntry={true}
      style={
        Array [
          Object {
            "backgroundColor": "#FFF",
            "borderColor": "#ECECEC",
            "borderRadius": 3,
            "borderWidth": 1,
            "padding": 16,
          },
          Object {
            "marginTop": 10,
          },
        ]
      }
      underlineColorAndroid="transparent"
    />
  </View>
  <View
    style={
      Object {
        "paddingTop": 16,
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#BBB",
          "marginBottom": 16,
        }
      }
    >
      Hey @undefined! It looks like you haven't setup your encrypted chat password yet. We recommend that you use a different password than your account password for added security.
    </Text>
  </View>
</View>
`;

exports[`Messenger setup component should render correctly for unlock 1`] = `
<View
  style={
    Array [
      Object {
        "flex": 1,
      },
      Object {
        "padding": 10,
      },
      Object {
        "backgroundColor": "#f7f7f7",
      },
    ]
  }
>
  <View
    style={
      Object {
        "alignItems": "stretch",
        "flexDirection": "column",
      }
    }
  >
    <TextInput
      allowFontScaling={true}
      editable={true}
      onChangeText={[Function]}
      placeholder="password..."
      secureTextEntry={true}
      style={
        Object {
          "backgroundColor": "#FFF",
          "borderColor": "#ECECEC",
          "borderRadius": 3,
          "borderWidth": 1,
          "padding": 16,
        }
      }
      underlineColorAndroid="transparent"
    />
  </View>
  <View
    style={
      Object {
        "paddingTop": 32,
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#BBB",
          "marginBottom": 16,
        }
      }
    >
      路 You only need to enter this encryption password once as long as you stay signed in.
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#BBB",
          "marginBottom": 16,
        }
      }
    >
      路 It is important so that no one other than you and the people you are communicating with can access the content of your messages.
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#BBB",
          "marginBottom": 16,
        }
      }
    >
      路 By default the content of your messages is fully encrypted. For heightened security you may wish to go to your app settings and turn off push notifications in order to disallow metadata from being tracked.
    </Text>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#BBB",
          "marginBottom": 16,
        }
      }
    >
      路 By default the content of your messages is fully encrypted. For heightened security you may wish to go to your app settings and turn off push notifications in order to disallow metadata from being tracked.
    </Text>
  </View>
</View>
`;
